Application/StorageManager/Implementations/CookieTuitionStorageManager.cs

```diff
@@ -0,0 +1,181 @@
+
+        return 1;
+    }
+
+    ///<inheritdoc/>>
+    public string? GetLocalAuthority()
+    {
+        var result = GetAllTuitionPartners().ToList();
+        if (result.Any()) _localAuthority = result.First().LocalAuthority;
+
+        return _localAuthority;
+    }
+
+    private bool IsTuitionPartnerValid(TuitionPartner tuitionPartner)
+    {
+        var errorCount = 0;
+        if (string.IsNullOrEmpty(tuitionPartner.SeoUrl))
+        {
+            errorCount += 1;
+            _logger.LogError($"An invalid {nameof(TuitionPartner)}.{nameof(tuitionPartner.SeoUrl)} was provided");
+        }
+
+        if (!string.IsNullOrEmpty(tuitionPartner.LocalAuthority)) return errorCount == 0;
+
+        _logger.LogError($"An invalid {nameof(TuitionPartner)}.{nameof(tuitionPartner.LocalAuthority)} was provided");
+
+        return false;
+    }
+
+    private string ConstructCookieKey(string key)
+    {
+        char[] brackets = { '(', '[', '{', '}', ']', ')' };
+        if (key.IndexOfAny(brackets) >= 0)
+            key = $"{ReplaceBracket(key)}{ContainsBracket}";
+
+        return $"{CookieKeyFormat}{key}";
+    }
+
+    private string ReplaceBracket(string value) => value.Replace("(", "400028")
+        .Replace(")", "410029").Replace("[", "91005B")
+        .Replace("]", "93005D").Replace("{", "123007B")
+        .Replace("}", "125007D");
+
+    private string PutBracketBack(string value) => value.Replace("400028", "(")
+        .Replace("410029", ")").Replace("91005B", "[")
+        .Replace("93005D", "]").Replace("123007B", "{")
+        .Replace("125007D", "}");
+
+    private bool IsTuitionPartnerPresent(string key)
+    {
+        var isPresent = _httpContextAccessor.HttpContext?.Request.Cookies.ContainsKey(key);
+
+        return isPresent != null && isPresent.Value;
+    }
+
+    private void RemoveCookie(string cookieKey) => _httpContextAccessor.HttpContext?.Response.Cookies.Delete(cookieKey);
+
+    private string StringifyTuitionPartnerDetail(TuitionPartner tuitionPartner) =>
+        $"{tuitionPartner.SeoUrl}&{tuitionPartner.LocalAuthority}";
+
+    private void AddTuitionPartnerToCookie(string cookieKey, string stringifyTpDetail)
+    {
+        var cookieOptions = new CookieOptions();
+        cookieOptions.Expires = DateTime.Now.AddDays(1);
+        _httpContextAccessor.HttpContext?.Response.Cookies.Append(cookieKey, stringifyTpDetail, cookieOptions);
+    }
+
+    private IEnumerable<KeyValuePair<string, string>>? GetAllTuitionPartnerCookies() =>
+        _httpContextAccessor.HttpContext?.Request.Cookies
+            .Where(c => c.Key.Contains(CookieKeyFormat));
+
+    private TuitionPartner? GetTuitionPartnerFromCookie(string key)
+    {
+        var cookieKey = ConstructCookieKey(key);
+        if (!IsTuitionPartnerPresent(cookieKey)) return null;
+
+        var cookie = _httpContextAccessor.HttpContext?.Request.Cookies[$"{cookieKey}"];
+        var tuitionPartnerDetails = cookie?.Split("&");
```

Comment:
Seems strange that we're having to handle url form encoded values ourselves. Could you use something like the `CookieHeaderValue` class, as per this https://learn.microsoft.com/en-us/aspnet/web-api/overview/advanced/http-cookies#structured-cookie-data ?

Reply:
I definitely agree with the 'CookieHeaderValue' implementation. After looking into it again in more detail. I currently can't find a way to make it work in a class project. When using it in a web API/MVC/Razor pages project, the CookieHeaderValue implementation seems happy. But in a class project, [even after using different NuGet packages] it doesn't seem to work.